File: ruby/find_first_duplicate.rb
require 'set'

def find_first_duplicate(arr)

  # type your code in here
  first_dupe = Set.new

  arr.each do |item|
    return item if first_dupe.include?(item)
    first_dupe.add(item)
  end

  -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  // type your code here
  const firstDupe = new Set()
  for (const value of arr) {
    if (firstDupe.has(value)) {
      return value
    }
    firstDupe.add(value)
  }
  return -1
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// And a written explanation of your solution
